File: PatronesCreacionales/Builder.java
package PatronesCreacionales;
import java.util.ArrayList;
import java.util.List;


class Part {
    private final String name;
    private final double price;

    public Part(String name, double price) {
        this.name = name;
        this.price = price;
    }
    public double getPrice() { return price; }
    @Override public String toString() {
        return String.format("%s: $%.2f", name, price);
    }
}


class AutoKit {
    private final List<Part> parts = new ArrayList<>();
    public void addPart(Part p) { parts.add(p); }
    public double getPrice() {
        return parts.stream().mapToDouble(Part::getPrice).sum();
    }
    @Override public String toString() {
        StringBuilder sb = new StringBuilder("Kit de Auto:\n");
        for (Part p : parts) sb.append(p).append("\n");
        sb.append(String.format("Precio total: $%.2f", getPrice()));
        return sb.toString();
    }
}


class AutoKitBuilder {
    private final AutoKit kit = new AutoKit();

    public AutoKitBuilder addBrakePads() {
        kit.addPart(new Part("Pastillas de freno (juego)", 35));
        return this;
    }
    public AutoKitBuilder addOilFilter() {
        kit.addPart(new Part("Filtro de aceite", 8));
        return this;
    }
    public AutoKitBuilder addEngineOil() {
        kit.addPart(new Part("Aceite de motor 4L", 28));
        return this;
    }
    public AutoKitBuilder addSparkPlugs(int qty) {
        kit.addPart(new Part("Bujías x" + qty, 5 * qty));
        return this;
    }
    public AutoKitBuilder addWipers() {
        kit.addPart(new Part("Plumillas (par)", 12));
        return this;
    }
    public AutoKit getKit() { return kit; }
}


public class Builder {
    public static void main(String[] args) {
        AutoKit kit = new AutoKitBuilder()
                .addOilFilter()
                .addEngineOil()
                .addBrakePads()
                .addSparkPlugs(4)
                .addWipers()
                .getKit();

        System.out.println(kit);
    }
}

File: PatronesCreacionales/builder.py
class Part:
    def __init__(self, name, price):
        self.name = name
        self.price = float(price)

    def get_price(self):
        return self.price

    def __str__(self):
        return f"{self.name}: ${self.price:.2f}"


class AutoKit:
    def __init__(self):
        self.parts = []

    def add_part(self, part):
        self.parts.append(part)

    def get_price(self):
        return sum(p.get_price() for p in self.parts)

    def __str__(self):
        lines = ["Kit de Auto:"]
        lines += [str(p) for p in self.parts]
        lines += [f"Precio total: ${self.get_price():.2f}"]
        return "\n".join(lines)


class AutoKitBuilder:
    def __init__(self):
        self.kit = AutoKit()

    # Componentes comunes
    def add_brake_pads(self):
        self.kit.add_part(Part("Pastillas de freno (juego)", 35))
        return self

    def add_oil_filter(self):
        self.kit.add_part(Part("Filtro de aceite", 8))
        return self

    def add_engine_oil(self):
        self.kit.add_part(Part("Aceite de motor 4L", 28))
        return self

    def add_spark_plugs(self, qty=4):
        self.kit.add_part(Part(f"Bujías x{qty}", 5 * qty))
        return self

    def add_wipers(self):
        self.kit.add_part(Part("Plumillas (par)", 12))
        return self

    def get_kit(self):
        return self.kit


# --- Uso ---
builder = AutoKitBuilder()
kit = (builder
       .add_oil_filter()
       .add_engine_oil()
       .add_brake_pads()
       .add_spark_plugs(qty=4)
       .add_wipers()
       .get_kit())

print(kit)
